File: src/g1_base.cpp
#include "g1_base.hpp"

#include <chrono>
#include <cmath>

#include <viam/sdk/log/logging.hpp>

namespace viam_unitree {

G1Base::G1Base(const viam::sdk::Dependencies& deps, const viam::sdk::ResourceConfig& cfg)
    : Base(cfg.name()) {
    auto attrs = cfg.attributes();
    std::string network_interface = "eth0";

    auto it = attrs.find("network_interface");
    if (it != attrs.end()) {
        if (auto* val = it->second.get<std::string>()) {
            network_interface = *val;
        }
    }

    VIAM_SDK_LOG(info) << "Initializing G1Base with network interface: " << network_interface;

    unitree::robot::ChannelFactory::Instance()->Init(0, network_interface);
    loco_client_.Init();
    loco_client_.SetTimeout(10.f);

    VIAM_SDK_LOG(info) << "G1Base initialized successfully";
}

G1Base::~G1Base() {
    cancel_movement();
    loco_client_.StopMove();
}

std::vector<std::string> G1Base::validate(const viam::sdk::ResourceConfig& cfg) {
    return {};
}

void G1Base::cancel_movement() {
    {
        std::lock_guard<std::mutex> lock(mu_);
        cancelled_ = true;
    }
    cancel_cv_.notify_all();
}

void G1Base::stop(const viam::sdk::ProtoStruct& extra) {
    cancel_movement();
    loco_client_.StopMove();
    moving_ = false;
}

void G1Base::move_straight(int64_t distance_mm,
                           double mm_per_sec,
                           const viam::sdk::ProtoStruct& extra) {
    if (distance_mm == 0 || mm_per_sec == 0) {
        return;
    }

    double speed_mps = std::abs(mm_per_sec) / 1000.0;
    double duration_sec = std::abs(static_cast<double>(distance_mm)) / std::abs(mm_per_sec);
    double direction = (distance_mm > 0) ? 1.0 : -1.0;
    float vx = static_cast<float>(direction * speed_mps);

    {
        std::lock_guard<std::mutex> lock(mu_);
        cancelled_ = false;
    }
    moving_ = true;

    auto end_time =
        std::chrono::steady_clock::now() + std::chrono::duration<double>(duration_sec);

    while (std::chrono::steady_clock::now() < end_time) {
        {
            std::unique_lock<std::mutex> lock(mu_);
            if (cancelled_) {
                moving_ = false;
                return;
            }
        }

        loco_client_.Move(vx, 0, 0);

        {
            std::unique_lock<std::mutex> lock(mu_);
            cancel_cv_.wait_for(lock, std::chrono::milliseconds(100), [this] {
                return cancelled_;
            });
            if (cancelled_) {
                moving_ = false;
                return;
            }
        }
    }

    loco_client_.StopMove();
    moving_ = false;
}

void G1Base::spin(double angle_deg,
                  double degs_per_sec,
                  const viam::sdk::ProtoStruct& extra) {
    if (angle_deg == 0 || degs_per_sec == 0) {
        return;
    }

    double duration_sec = std::abs(angle_deg) / std::abs(degs_per_sec);
    double direction = (angle_deg > 0) ? 1.0 : -1.0;
    float vyaw = static_cast<float>(direction * std::abs(degs_per_sec) * M_PI / 180.0);

    {
        std::lock_guard<std::mutex> lock(mu_);
        cancelled_ = false;
    }
    moving_ = true;

    auto end_time =
        std::chrono::steady_clock::now() + std::chrono::duration<double>(duration_sec);

    while (std::chrono::steady_clock::now() < end_time) {
        {
            std::unique_lock<std::mutex> lock(mu_);
            if (cancelled_) {
                moving_ = false;
                return;
            }
        }

        loco_client_.Move(0, 0, vyaw);

        {
            std::unique_lock<std::mutex> lock(mu_);
            cancel_cv_.wait_for(lock, std::chrono::milliseconds(100), [this] {
                return cancelled_;
            });
            if (cancelled_) {
                moving_ = false;
                return;
            }
        }
    }

    loco_client_.StopMove();
    moving_ = false;
}

void G1Base::set_power(const viam::sdk::Vector3& linear,
                       const viam::sdk::Vector3& angular,
                       const viam::sdk::ProtoStruct& extra) {
    // Map power (-1 to 1) to reasonable velocities for G1.
    // G1 max forward speed ~1.5 m/s, max rotation ~1.0 rad/s.
    constexpr float kMaxLinearVel = 1.5f;
    constexpr float kMaxAngularVel = 1.0f;

    float vx = static_cast<float>(linear.x) * kMaxLinearVel;
    float vy = static_cast<float>(linear.y) * kMaxLinearVel;
    float vyaw = static_cast<float>(angular.z) * kMaxAngularVel;

    moving_ = (vx != 0 || vy != 0 || vyaw != 0);
    loco_client_.Move(vx, vy, vyaw);
}

void G1Base::set_velocity(const viam::sdk::Vector3& linear,
                          const viam::sdk::Vector3& angular,
                          const viam::sdk::ProtoStruct& extra) {
    // Viam: linear in mm/s, angular in deg/s.
    // Unitree: linear in m/s, angular in rad/s.
    float vx = static_cast<float>(linear.x / 1000.0);
    float vy = static_cast<float>(linear.y / 1000.0);
    float vyaw = static_cast<float>(angular.z * M_PI / 180.0);

    moving_ = (vx != 0 || vy != 0 || vyaw != 0);
    loco_client_.Move(vx, vy, vyaw);
}

bool G1Base::is_moving() {
    return moving_;
}

viam::sdk::ProtoStruct G1Base::get_status() {
    viam::sdk::ProtoStruct status;
    status["is_moving"] = moving_.load();
    return status;
}

viam::sdk::Base::properties G1Base::get_properties(const viam::sdk::ProtoStruct& extra) {
    // G1 humanoid: ~0.45m wide, not wheeled.
    return {0.45, 0, 0};
}

viam::sdk::ProtoStruct G1Base::do_command(const viam::sdk::ProtoStruct& command) {
    auto it = command.find("command");
    if (it == command.end()) {
        return {};
    }

    auto* cmd = it->second.get<std::string>();
    if (!cmd) {
        return {};
    }

    viam::sdk::ProtoStruct result;
    int32_t rc = 0;

    if (*cmd == "stand_up") {
        rc = loco_client_.StandUp();
    } else if (*cmd == "sit") {
        rc = loco_client_.Sit();
    } else if (*cmd == "squat") {
        rc = loco_client_.Squat();
    } else if (*cmd == "high_stand") {
        rc = loco_client_.HighStand();
    } else if (*cmd == "low_stand") {
        rc = loco_client_.LowStand();
    } else if (*cmd == "balance_stand") {
        rc = loco_client_.BalanceStand();
    } else if (*cmd == "damp") {
        rc = loco_client_.Damp();
    } else if (*cmd == "zero_torque") {
        rc = loco_client_.ZeroTorque();
    } else if (*cmd == "wave_hand") {
        rc = loco_client_.WaveHand();
    } else if (*cmd == "start") {
        rc = loco_client_.Start();
    } else if (*cmd == "stop_move") {
        rc = loco_client_.StopMove();
    } else {
        result["error"] = std::string("unknown command: " + *cmd);
        return result;
    }

    result["rc"] = static_cast<double>(rc);
    return result;
}

std::vector<viam::sdk::GeometryConfig> G1Base::get_geometries(
    const viam::sdk::ProtoStruct& extra) {
    return {};
}

}  // namespace viam_unitree

File: g1_base.go
package main

import (
	"context"
	"fmt"
	"math"
	"sync"
	"sync/atomic"
	"time"

	"github.com/golang/geo/r3"

	"go.viam.com/rdk/components/base"
	"go.viam.com/rdk/logging"
	"go.viam.com/rdk/resource"
	"go.viam.com/rdk/spatialmath"
)

var g1BaseModel = resource.NewModel("erh", "viam-unitree", "g1-base")

type G1BaseConfig struct {
	NetworkInterface string `json:"network_interface"`
}

func (c *G1BaseConfig) Validate(path string) ([]string, error) {
	return nil, nil
}

func init() {
	resource.RegisterComponent(base.API, g1BaseModel, resource.Registration[base.Base, *G1BaseConfig]{
		Constructor: newG1Base,
	})
}

type g1Base struct {
	resource.Named
	resource.AlwaysRebuild

	logger logging.Logger
	loco   *LocoClient

	mu        sync.Mutex
	moving    atomic.Bool
	cancelCtx context.Context
	cancelFn  context.CancelFunc
}

func newG1Base(ctx context.Context, deps resource.Dependencies, conf resource.Config, logger logging.Logger) (base.Base, error) {
	cfg, err := resource.NativeConfig[*G1BaseConfig](conf)
	if err != nil {
		return nil, err
	}

	networkInterface := "eth0"
	if cfg.NetworkInterface != "" {
		networkInterface = cfg.NetworkInterface
	}

	logger.Infof("Initializing G1Base with network interface: %s", networkInterface)

	if err := InitChannel(0, networkInterface); err != nil {
		return nil, fmt.Errorf("channel init: %w", err)
	}

	loco, err := NewLocoClient()
	if err != nil {
		return nil, fmt.Errorf("loco client create: %w", err)
	}
	if err := loco.Init(); err != nil {
		loco.Close()
		return nil, fmt.Errorf("loco client init: %w", err)
	}
	loco.SetTimeout(10.0)

	cancelCtx, cancelFn := context.WithCancel(context.Background())

	logger.Info("G1Base initialized successfully")

	return &g1Base{
		Named:     conf.ResourceName().AsNamed(),
		logger:    logger,
		loco:      loco,
		cancelCtx: cancelCtx,
		cancelFn:  cancelFn,
	}, nil
}

func (b *g1Base) MoveStraight(ctx context.Context, distanceMm int, mmPerSec float64, extra map[string]interface{}) error {
	if distanceMm == 0 || mmPerSec == 0 {
		return nil
	}

	speedMps := math.Abs(mmPerSec) / 1000.0
	durationSec := math.Abs(float64(distanceMm)) / math.Abs(mmPerSec)
	direction := 1.0
	if distanceMm < 0 {
		direction = -1.0
	}
	vx := float32(direction * speedMps)

	b.mu.Lock()
	// Create a new cancel context for this movement.
	b.cancelFn()
	moveCtx, moveFn := context.WithCancel(context.Background())
	b.cancelCtx = moveCtx
	b.cancelFn = moveFn
	b.mu.Unlock()

	b.moving.Store(true)
	defer b.moving.Store(false)

	deadline := time.Now().Add(time.Duration(durationSec * float64(time.Second)))
	ticker := time.NewTicker(100 * time.Millisecond)
	defer ticker.Stop()

	for time.Now().Before(deadline) {
		select {
		case <-ctx.Done():
			b.loco.StopMove()
			return ctx.Err()
		case <-moveCtx.Done():
			return nil
		case <-ticker.C:
			b.loco.Move(vx, 0, 0)
		}
	}

	b.loco.StopMove()
	return nil
}

func (b *g1Base) Spin(ctx context.Context, angleDeg, degsPerSec float64, extra map[string]interface{}) error {
	if angleDeg == 0 || degsPerSec == 0 {
		return nil
	}

	durationSec := math.Abs(angleDeg) / math.Abs(degsPerSec)
	direction := 1.0
	if angleDeg < 0 {
		direction = -1.0
	}
	vyaw := float32(direction * math.Abs(degsPerSec) * math.Pi / 180.0)

	b.mu.Lock()
	b.cancelFn()
	moveCtx, moveFn := context.WithCancel(context.Background())
	b.cancelCtx = moveCtx
	b.cancelFn = moveFn
	b.mu.Unlock()

	b.moving.Store(true)
	defer b.moving.Store(false)

	deadline := time.Now().Add(time.Duration(durationSec * float64(time.Second)))
	ticker := time.NewTicker(100 * time.Millisecond)
	defer ticker.Stop()

	for time.Now().Before(deadline) {
		select {
		case <-ctx.Done():
			b.loco.StopMove()
			return ctx.Err()
		case <-moveCtx.Done():
			return nil
		case <-ticker.C:
			b.loco.Move(0, 0, vyaw)
		}
	}

	b.loco.StopMove()
	return nil
}

func (b *g1Base) SetPower(ctx context.Context, linear, angular r3.Vector, extra map[string]interface{}) error {
	const maxLinearVel = 1.5
	const maxAngularVel = 1.0

	vx := float32(linear.X * maxLinearVel)
	vy := float32(linear.Y * maxLinearVel)
	vyaw := float32(angular.Z * maxAngularVel)

	b.moving.Store(vx != 0 || vy != 0 || vyaw != 0)
	b.loco.Move(vx, vy, vyaw)
	return nil
}

func (b *g1Base) SetVelocity(ctx context.Context, linear, angular r3.Vector, extra map[string]interface{}) error {
	// Viam: linear in mm/s, angular in deg/s.
	// Unitree: linear in m/s, angular in rad/s.
	vx := float32(linear.X / 1000.0)
	vy := float32(linear.Y / 1000.0)
	vyaw := float32(angular.Z * math.Pi / 180.0)

	b.moving.Store(vx != 0 || vy != 0 || vyaw != 0)
	b.loco.Move(vx, vy, vyaw)
	return nil
}

func (b *g1Base) Stop(ctx context.Context, extra map[string]interface{}) error {
	b.mu.Lock()
	b.cancelFn()
	b.mu.Unlock()

	b.loco.StopMove()
	b.moving.Store(false)
	return nil
}

func (b *g1Base) IsMoving(ctx context.Context) (bool, error) {
	return b.moving.Load(), nil
}

func (b *g1Base) Properties(ctx context.Context, extra map[string]interface{}) (base.Properties, error) {
	return base.Properties{
		WidthMeters: 0.45,
	}, nil
}

func (b *g1Base) Geometries(ctx context.Context, extra map[string]interface{}) ([]spatialmath.Geometry, error) {
	return nil, nil
}

func (b *g1Base) DoCommand(ctx context.Context, cmd map[string]interface{}) (map[string]interface{}, error) {
	cmdStr, ok := cmd["command"].(string)
	if !ok {
		return map[string]interface{}{}, nil
	}

	var rc int
	switch cmdStr {
	case "stand_up":
		rc = b.loco.StandUp()
	case "sit":
		rc = b.loco.Sit()
	case "squat":
		rc = b.loco.Squat()
	case "high_stand":
		rc = b.loco.HighStand()
	case "low_stand":
		rc = b.loco.LowStand()
	case "balance_stand":
		rc = b.loco.BalanceStand()
	case "damp":
		rc = b.loco.Damp()
	case "zero_torque":
		rc = b.loco.ZeroTorque()
	case "wave_hand":
		rc = b.loco.WaveHand()
	case "start":
		rc = b.loco.Start()
	case "stop_move":
		rc = b.loco.StopMove()
	default:
		return map[string]interface{}{"error": "unknown command: " + cmdStr}, nil
	}

	return map[string]interface{}{"rc": float64(rc)}, nil
}

func (b *g1Base) Close(ctx context.Context) error {
	b.mu.Lock()
	b.cancelFn()
	b.mu.Unlock()

	b.loco.StopMove()
	b.loco.Close()
	return nil
}
